solarforecastarbiter/metrics/tests/test_calculator.py

```diff
@@ -296,3 +297,91 @@ def test_apply_deterministic_bad_metric_func():
         calculator._apply_deterministic_metric_func('BAD METRIC',
                                                     pd.Series(),
                                                     pd.Series())
+
+
+@pytest.mark.parametrize('categories,metrics', [
+    ([], []),
+    (LIST_OF_CATEGORIES, DETERMINISTIC_METRICS)
+])
+def test_interval_label(
+    site_metadata, categories, metrics, create_processed_fxobs
+):
+    ts = pd.date_range(
+        start="2019-07-01 00:00:00",
+        end="2019-07-08 00:00:00",
+        freq="1h",
+        tz="MST",
+        name="timestamp"
+    )
+    fx = pd.Series(data=np.random.rand(len(ts)), index=ts, name="value")
+    obs = pd.Series(data=np.random.rand(len(ts)), index=ts, name="value")
+    ref = pd.Series(data=np.random.rand(len(ts)), index=ts, name="value")
+
+    # `interval_label == beginning`
+    fx_obs = datamodel.ForecastObservation(
+        forecast=datamodel.Forecast(
+            site=site_metadata,
+            name="dummy fx",
+            variable="ghi",
+            interval_value_type="instantaneous",
+            interval_length=pd.Timedelta(fx.index.freq),
+            interval_label="ending",
+            issue_time_of_day=datetime.time(hour=5),
+            lead_time_to_start=pd.Timedelta('1h'),
+            run_length=pd.Timedelta('12h')
+        ),
+        observation=datamodel.Observation(
+            site=site_metadata,
+            name="dummy obs",
+            variable="ghi",
+            interval_value_type="instantaneous",
+            interval_length=pd.Timedelta(obs.index.freq),
+            interval_label="ending",
+            uncertainty=1,
+        )
+    )
+
+    # `interval_label == ending`
+    ref_obs = datamodel.ForecastObservation(
+        forecast=datamodel.Forecast(
+            site=site_metadata,
+            name="dummy fx",
+            variable="ghi",
+            interval_value_type="instantaneous",
+            interval_length=pd.Timedelta(fx.index.freq),
+            interval_label="beginning",
+            issue_time_of_day=datetime.time(hour=5),
+            lead_time_to_start=pd.Timedelta('1h'),
+            run_length=pd.Timedelta('12h')
+        ),
+        observation=datamodel.Observation(
+            site=site_metadata,
+            name="dummy obs",
+            variable="ghi",
+            interval_value_type="instantaneous",
+            interval_length=pd.Timedelta(obs.index.freq),
+            interval_label="beginning",
+            uncertainty=1,
+        )
+    )
+
+    proc_fx_obs = create_processed_fxobs(fx_obs, fx, obs)
+    proc_ref_obs = create_processed_fxobs(ref_obs, ref, obs)
+
+    if any(m in deterministic._REQ_REF_FX for m in metrics):
```

Comment:
what is this line supposed to do?

Reply:
I copied that line from `test_calculate_metrics()` as a way to iterate through the set of (user-selected) metrics for only those metrics where you need a reference forecast.

Reply:
I'm still confused. Is it meant to only run the test code for the non-empty lists in the parametrization? Do we actually need that here?

I was trying to understand why the CI is still failing, and got caught on this first.

Reply:
Correction: that exact line is taken from `metrics.calculator.calculate_deterministic_metrics()`; a similar line is seen in `test_calculator_metrics() in `test_calculator.py`.

Reply:
> I'm still confused. Is it meant to only run the test code for the non-empty lists in the parametrization? Do we actually need that here?
> 
> I was trying to understand why the CI is still failing, and got caught on this first.

Let me double-check that it's working as I meant, but my goal was to run the `pytest.raises(ValueError)` test only on metrics that require a reference forecast.

Reply:
I know other tests do this, but I think we want to move away from if statements to hit different paths in a function. I prefer making multiple tests instead. I.e the standard test could be ``test_interval_label``, and you might have another test ``test_interval_label_with_reference_forecast`` that would test for the ValueError. You can use pytest fixtures to hold any boilerplate and sometimes using ``pytest.param(..., marks=pytest.mark.xfail(...))`` in the parametrize is sufficient

Reply:
Multiple (simpler/explicit) tests makes sense to me. I'll revise the current (single) test to separate tests for the `interval_label == "ending"` `groupby()` behavior and another for the mistmatched `interval_label` scenario.